index.bs

```diff
@@ -795,6 +743,562 @@ the following list:
+
+ 1. If [=WebSocket/ready state=] is not {{WebSocket/OPEN}} (1), then return.
+ 1. Let |chunk| be determined by switching on |type|:
+
+  <dl class="switch">
+   : |type| indicates that the data is Text
+   :: a new {{DOMString}} containing |data|
+
+   : |type| indicates that the data is Binary
+   :: a new {{Uint8Array}} object, created in the [=relevant Realm=] of the {{WebSocketStream}}
+      object, whose contents are |data|
+  </dl>
+
+ 1. [=ReadableStream/Enqueue=] |chunk| into |stream|'s [=WebSocketStream/readable stream=].
+ 1. Apply backpressure to the WebSocket.
+
+ Note: Applying backpressure <span class=allow-2119>should</span> result in no new WebSocket
+ messages being handled until the backpressure is released. After an implementation-defined amount
+ of data is buffered, the client will stop reading from the underlying network connection resulting
+ in network-layer backpressure being applied.
+</div>
+
+<hr>
+
+When [=the WebSocket closing handshake is started=] for a {{WebSocketStream}} |stream|, the user
+agent must [=queue a task=] to change the |stream|'s [=WebSocketStream/ready state=] to
+{{WebSocket/CLOSING}} if it is not already. [[!WSP]]
+
+<hr>
+
+<div algorithm="handle WebSocketStream connection close">
+
+When [=the WebSocket connection is closed=] for a {{WebSocketStream}} |stream|, possibly
+ [=cleanly=], the user agent must [=queue a task=] to run the following substeps:
+
+ 1. Change the [=WebSocketStream/ready state=] to {{WebSocket/CLOSED}} (3).
+ 1. If |stream|'s [=WebSocketStream/handshake aborted flag=] is set, then return.
+ 1. If |stream|'s [=WebSocketStream/was ever connected flag=] is not set, then [=reject=] |stream|'s
+    [=opened promise=] with a [=new=] WebSocketError.
+ 1. Let |code| be [=the WebSocket connection close code=].
+ 1. Let |reason| be the result of applying [=UTF-8 decode without BOM=] to [=the WebSocket
+    connection close reason=].
+ 1. If the connection was closed [=cleanly=],
+  1. [=ReadableStream/Close=] |stream|'s [=WebSocketStream/readable stream=].
+  1. [=WritableStream/Error=] |stream|'s [=WebSocketStream/writable stream=] with an
+     "{{InvalidStateError}}" {{DOMException}} indicating that a closed {{WebSocketStream}} cannot be
+     written to.
+  1. [=Resolve=] |stream|'s [=closed promise=] with a [=new=] {{WebSocketCloseInfo}} constructed
+     with |code| and |reason|.
+ 1. Otherwise,
+  1. [=ReadableStream/Error=] |stream|'s [=WebSocketStream/readable stream=] with a [=new=]
+     {{WebSocketError}} constructed with |code| and |reason|.
+  1. [=WritableStream/Error=] |stream|'s [=WebSocketStream/writable stream=] with a [=new=]
+     {{WebSocketError}} constructed with |code| and |reason|.
+  1. [=Reject=] |stream|'s [=closed promise=] with a [=new=] {{WebSocketError}} constructed with
+     |code| and |reason|.
+</div>
+
+<hr>
+
+The [=task source=] for all [=tasks=] <a lt="queue a task">queued</a> in this section is the
+[=WebSocket task source=].
```

Comment:
Ah, I see the task source. I don't like the action at a distance here; we've tried to move away from it in the past.

Reply:
I've removed this bit in favour of explicitly specifying it each time.